### PR TITLE
gitlab-shell: 8.3.3->8.4.1, fix hardcoded paths

### PR DESCRIPTION
**Pull request by krav**
###### Motivation for this change

This upgrades `gitlab-shell` to be current with `gitlab`.

In addition it fixes paths in the generated `authorized_keys` file, which hardcoded paths to the running gitlab-shell and broke upgrades.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x]Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/applications/version-management/gitlab-shell/remove-hardcoded-locations.patch [hunk 1]
```diff
@@ -1,23 +1,33 @@
...
- 	}
-diff --git a/lib/gitlab_shell.rb b/lib/gitlab_shell.rb
-index 1452f95..2b40327 100644
---- a/lib/gitlab_shell.rb
-+++ b/lib/gitlab_shell.rb
-@@ -180,7 +180,8 @@ class GitlabShell
-     end
+ func NewFromDir(dir string) (*Config, error) {
+-	return newFromFile(path.Join(dir, configFile))
++	return newFromFile("/var/run/gitlab/shell-config.yml")
```
Comment (Mic92) [inline pkgs/applications/version-management/gitlab-shell/remove-hardcoded-locations.patch]: Should this be `/run` without `/var`?

**Comment by globin** [comment]
Per GITLAB_SHELL_VERSION in the latest gitlab releases it should be 8.3.3

File: pkgs/applications/version-management/gitlab-shell/remove-hardcoded-locations.patch (hunk 1 above)
Comment (krav) [reply pkgs/applications/version-management/gitlab-shell/remove-hardcoded-locations.patch]: Probably. /var/run is a symlink to /run on my system, didn't notice that before.

**Comment by krav** [comment]
> Per GITLAB_SHELL_VERSION in the latest gitlab releases it should be 8.3.3
Oops, seems I accidentally looked at the master branch. Should I change this PR to include only the patch? Or just let it sit here until the next release the 22.?

GitLab will break for existing users after upgrading and garbage collecting, since `/var/gitlab/state/home/.ssh/authorized_keys` is not regenerated. Is there a place to inform users of this, or should we fix it for them with `ExecStarPre=` in the unit? Not a big fan of the latter, because it's not clear how long such a thing should persist.

File: pkgs/applications/version-management/gitlab-shell/remove-hardcoded-locations.patch [hunk 2]
```diff
@@ -1,23 +1,33 @@
...
+ func newFromFile(filename string) (*Config, error) {
+diff -Nuar gitlab-shell.orig/lib/gitlab_keys.rb gitlab-shell/lib/gitlab_keys.rb
+--- gitlab-shell.orig/lib/gitlab_keys.rb	2018-10-29 11:18:37.527010419 +0100
++++ gitlab-shell/lib/gitlab_keys.rb	2018-10-29 11:29:39.462454083 +0100
+@@ -10,7 +10,7 @@
+   attr_accessor :auth_file, :key
+ 
+   def self.command(whatever)
+-    "#{ROOT_PATH}/bin/gitlab-shell #{whatever}"
++    "/var/run/current-system/sw/bin/gitlab-shell #{whatever}"
```
Comment (Mic92) [inline pkgs/applications/version-management/gitlab-shell/remove-hardcoded-locations.patch]: Should also point to `/run/`

File: pkgs/applications/version-management/gitlab-shell/remove-hardcoded-locations.patch [hunk 3]
```diff
@@ -1,23 +1,33 @@
...
- 	}
-diff --git a/lib/gitlab_shell.rb b/lib/gitlab_shell.rb
-index 1452f95..2b40327 100644
---- a/lib/gitlab_shell.rb
-+++ b/lib/gitlab_shell.rb
-@@ -180,7 +180,8 @@ class GitlabShell
-     end
+ func NewFromDir(dir string) (*Config, error) {
+-	return newFromFile(path.Join(dir, configFile))
++	return newFromFile("/run/gitlab/shell-config.yml")
```
Comment (Mic92) [inline pkgs/applications/version-management/gitlab-shell/remove-hardcoded-locations.patch]: What is `dir` usually?

Comment (krav) [reply pkgs/applications/version-management/gitlab-shell/remove-hardcoded-locations.patch]: `os.Getwd()`, which I assume is the `gitlab` homedir, `/var/gitlab/state/home`. Note that this part of the patch is already included in nixpkgs.

**Comment by globin** [comment]
Please only include the patch, it's easy enough to bump gitlab shell when updating the rest

**Comment by flokli** [comment]
@krav  GitLab was bumped to 11.4.4, but gitlab-shell should still be `8.3.3`.

So as @globin said, please remove the bump to 8.4.1 for now.

**Comment by flokli** [comment]
@globin can you verify this works, and we probably can strip
```
        # The gitlab:shell:setup regenerates the authorized_keys file so that
        # the store path to the gitlab-shell in it gets updated
        ${pkgs.sudo}/bin/sudo -u ${cfg.user} -H force=yes ${gitlab-rake}/bin/gitlab-rake gitlab:shell:setup
```

from `nixos/modules/services/misc/gitlab.nix` too?

**Comment by flokli** [comment]
@krav Is there anything missing from here, after https://github.com/NixOS/nixpkgs/pull/50962 being merged? Can you rebase this on top of master?

**Comment by krav** [comment]
@flokli Done. I don't think there's anything missing, this is a reasonably simple change.

**Comment by flokli** [comment]
I squashed together the two commits changing `remove-hardcoded-locations.patch`, so it's more understandable that there's only a chunk patching `lib/gitlab_keys.rb` added.

**Comment by flokli** [comment]
@krav I stripped the `authorized_keys` file regeneration on each gitlab startup, and added a note about that to the changelog. I'm thinking about simply not backporting 658c865 to 18.09, so for 19.03 everybody will most likely already have a `authorized_keys` file that's using `/var/run/current-system/sw/bin/gitlab-shell` command :-)

File: nixos/doc/manual/release-notes/rl-1903.xml [hunk 4]
```diff
@@ -241,8 +241,22 @@
...
+  </listitem>
+  <listitem>
+   <para>
+     GitLab Shell previously used the nix store paths for the
+     <literal>gitlab-shell</literal> command in its
+     <literal>authorized_keys</literal> file, which might stop working after
+     garbage collection. To circumvent that, we regenerated that file on each
+     startup.  As <literal>gitlab-shell</literal> has now been changed to use
+     <literal>/var/run/current-system/sw/bin/gitlab-shell</literal>, this is
+     not necesary anymore, but there might be leftover lines with a nix store
```
Comment (Mic92) [inline nixos/doc/manual/release-notes/rl-1903.xml]:
```suggestion
     not necessary anymore, but there might be leftover lines with a nix store
```

**Comment by krav** [comment]
@flokli I have to admit I wasn't aware of the `rake` task. It's a recent addition to the module, it seems @WilliButz and I were solving the same problem. Not backporting 658c865 sounds sensible. I've looked through the relevant gitlab rake code, and it seems to fork() into gitlab-shell to regenerate the keys, so with this patch everyone should get `/run/current-system/` on restart.

**Comment by flokli** [comment]
Squashed together be77bbd and 62e5de4 into a3ec5dce2bcc7b346acd7a507821dc2bc105fbac for easier backporting into https://github.com/NixOS/nixpkgs/pull/51072

**Pull request merged**